eventmesh/LockManager.js

```diff
@@ -1,105 +1,184 @@
 'use strict';
 
 const _ = require('lodash');
-const config = require('../common/config');
-const logger = require('../common/logger');
+const Promise = require('bluebird');
+const eventmesh = require('./');
 const errors = require('../common/errors');
-const ETCDLockError = errors.ETCDLockError;
-const CONST = require('./constants');
-
-const {
-  Etcd3
-} = require('etcd3');
-const client = new Etcd3({
-  hosts: config.etcd.url,
-  credentials: {
-    rootCertificate: Buffer.from(config.etcd.ssl.ca, 'utf8'),
-    privateKey: Buffer.from(config.etcd.ssl.key, 'utf8'),
-    certChain: Buffer.from(config.etcd.ssl.crt, 'utf8')
-  }
-});
+const DeploymentAlreadyLocked = errors.DeploymentAlreadyLocked;
+const logger = require('../common/logger');
+const CONST = require('../common/constants');
+const NotFound = errors.NotFound;
+const Conflict = errors.Conflict;
+const InternalServerError = errors.InternalServerError;
 
 class LockManager {
   /*
-  Lock is tracked via the key value "resource/lock/details". 
-  The value is a JSON type and the structure looks like the following.
-  {count: INT, operationType:STRING}
-  when a lock is taken, count value is set to 1.
-  operationType value is either "READ or "WRITE", depending on if the ongoing operation will make sync operations like bind/unbind wait for it.
+  This method checks whether lock is of write type.
+  returns true if lock is present and not expired and it's of type WRITE
+  */
+  /**
+   * @param {string} resourceId - Id (name) of the resource that is being locked. In case of deployment lock, it is instance_id
+   */
 
-  This value is read and updated to acquire lock. In order to make this two different operations synchronous, we do it taking a lock on the "resource/lock" key, which would ensure synchrounous execution of the read and update operation.
+  isWriteLocked(resourceId) {
+    return eventmesh.apiServerClient.getLockDetails(CONST.APISERVER.RESOURCE_TYPES.DEPLOYMENT_LOCKS, resourceId)
+      .then(lockDetails => {
+        const currentTime = new Date();
+        const lockTime = new Date(lockDetails.lockTime);
+        const lockTTL = lockDetails.lockTTL ? lockDetails.lockTTL : Infinity;
+        if (lockDetails.lockType === CONST.ETCD.LOCK_TYPE.WRITE && ((currentTime - lockTime) < lockTTL)) {
+          logger.info(`Resource ${resourceId} was write locked for ${lockDetails.lockedResourceDetails.operation ?
+            lockDetails.lockedResourceDetails.operation : 'unknown'} ` +
+            `operation with id ${lockDetails.lockedResourceDetails.resourceId} at ${lockTime} `);
+          return true;
+        }
+        return false;
+      })
+      .catch(NotFound, () => false);
```

Comment:
It would be a good idea to return the lock details in case of locked, as it can be used to populate the DeploymentAlreadyLockedError with appropriate details.